### ISSUE TITLE
Rewrite runloop

### ISSUE THREAD

**Issue by erikvanoosten** [zio/zio-kafka#706]
Zio-kafka has a very nice programming model but has a few design choices that make it only suitable for high throughput use cases where duplicating or losing a few messages is not an issue.

This issue tries to summarize the problems and gives a rough design for a new runloop that will fix these problems.

# Problems

~*Problem 1 - commits can be arbitrarily delayed*~

Currently commits are delayed during a rebalance. However, there is no good way to know when a rebalance is happening. Unfortunately, using the rebalance listener for that is not reliable. For example when the only change in the rebalance is that a partition is revoked. In this case we do not see the end of a rebalance because only onRevoke will be called, and onAssigned will not be.

**Update:** ✅ With #683 commits are no longer blocked by a rebalance. Instead, they are directly sent to the underlying Kafka client.

*Problem 2 - message duplication after rebalance*

A clean rebalance does not lead to message duplication when a revoked partition gets the chance to finish its work and commit everything before the rebalance starts. This is the purpose of the onRevoked callback of the rebalance listener. Currently, we do not wait inside the callback until the streams that handle the revoked partitions are complete.

**Update:** This is implemented in https://github.com/zio/zio-kafka/pull/788.

~*Problem 3 - polling delays*~

Polling frequency is determined by the poll delay parameter. Better latency and throughput is achieved when poll is called continuously. An experiment by @svroonland (see #683) showed that calling poll continuously gives a ridiculous improvement in throughput.

**Update:** ✅ #683 which implemented this idea has been merged!

*Problem 4 - transactional producer is not well supported*

It is currently quite difficult to use a transactional producer with zio-kafka. It requires the use of a custom rebalance listener and the _restartStreamOnRebalancing_ mode. With the design proposed in this issue working with transactional producers becomes as easy as regular producers.

Another problem is that when a stream needs to be restarted, that the new stream can already start processing while the old stream is still processing older messages of the same partition. That means that all ordering guarantees are off and that commits might be executed in the wrong order (which can lead to duplicates).

# Related issues

This issue is a duplicate of/is related with:

#590: duplicates after rebalance
#664 and #428: optimize polling behavior
#479 is no longer needed since we'll only close the correct streams

What follows here is a rough design for a new run loop.

--------------------------

# Runloop

_NOTE: This document describes an alternative Runloop implementation, not the currently implemented Runloop_

**Update:** the poll loop model (continuous polling) and the stream control model (feed via a queue instead of a promise) as described below have been adopted in zio-kafka 2.2.

Runloop is the heart of zio-kafka.
It manages the underlying Java based Kafka client in such a way that messages can be safely processed in a concurrent fashion, using the convenient API of ZIO streams.

In this section we assume:
 - we are using subscribe (not assign),
 - we use one of the default partitioners (manual partition assignment is supported, but described in a separate section).

Runloop has the following tasks:
- Poll the underlying Java Kafka client for new events.
- Handle rebalance events.
- Create new streams as new partitions are assigned, remove streams as partitions are revoked or lost.
- Pause partitions for which there is no demand.
- Pass commits to the underlying Java Kafka client.
- Coordinate streams such that messages are not processed twice (where possible) and transactional generations do not overlap.

Runloop has these distinct run phases:
- poll loop,
- commit callbacks,
- rebalance listener.

The _poll loop_ does the following (in a loop):
- Collect all data requests from the streams.
- Pause partitions for which there is no request, unpause partitions for which there are requests.
- Call poll on the underlying Kafka client (note, this is a blocking call).
  - Meanwhile, commits complete and their callbacks will be invoked.
  - Meanwhile, the rebalance listener might be invoked.
- Forward stream commits to the underlying Kafka client.
- Pass new Kafka records to the streams.

The _commit callbacks_ are invoked when a commit completes.

The _rebalance listener_ does the following (when callback is invoked):
- upon revoked partitions:
  - handle incoming commit requests for the associated streams until they have finished
    (we only handle commits requests for the associated streams to prevent commit timeouts, other commits are stored in the runloop state)
    (to prevent duplicates, these commits must complete before leaving the callback)
    (see below about the maximum wait time),
- upon newly assigned partitions:
  - create a new stream for each partition,
- upon lost partitions:
  - mark the stream (or remove it from the state) so that from now on all data and commit requests from this stream will be rejected,
  - interrupt the existing stream.

Note that the commit callbacks and the rebalance listener are invoked from the same thread the poll loop is running on.
Therefore, all have thread-safe access to the same state.

To prevent being marked as an inactive client, poll must be called every `max.poll.interval.ms` (defaults to 5 min).
To make this possible, the rebalancer callbacks must not take too long.
We estimate that `max.poll.interval.ms / 2` is a good upper bound.

To make sure that the broker learns about a stalled client, the poll loop can stop calling poll on purpose.
The poll loop will only call poll when:
 - there is at least 1 pending request for data, or
 - there is at least 1 pending commit, or
 - a partition was revoked or lost in the last 5 polls.

When poll is not called, instead we wait on new command or commit messages.
(See also #683 where this idea is also being implemented.)
(See also #686 for the case when only a few streams are stalling.)

Runloop makes the following queues available for async communication:
 - command queue, for requests for data,
 - commit queue, for commit requests.

Note, these queues need to be separate because commits are also handled from the rebalance listener.

When a partition is assigned, it creates a stream.
Each stream has the following properties:
 - The topic and partition is handles.
 - A promise that is fulfilled when the stream completes.
 - A queue with chunks of received Kafka records.

A stream can be in 2 states: data-available (the initial state) or no-data-available.
In the data-available state the stream reads all available data from the queue without waiting.
Any read data is emitted.
As soon as there is no more data, a request for data is created and passed to the runloop.
The stream is now in the no-data-available state.
Next, it will do a blocking wait on the queue.
When more data becomes available it is emitted, and we move back to the data-available state.

## Transactional

Transactional producers require more attention than regular producers.
When a rebalance is started, a new transaction epoch is started as well.
We must take care that a producer does not commit anything from a newer epoch while it is still processing messages from the previous epoch.
In other words, we need to control the order of commits over different streams.
The most straightforward way to control this is by stopping all streams and restarting them after the rebalance completes.
This has some problems:
 - we stop all streams so throughput is affected,
 - when we do not get assigned new partitions, we do not know when the rebalance is completed.

A simpler approach is to use _one producer per partition_ (or more accurately: one `transactional.id` per partition).
(This approach is copied from the kafka-streams library.)

We need to change the API for this.
E.g. we need to get a `TransactionalProducerConfig` and create producers on the fly.
Optionally, we put the producers in a cache (the spring-kafka library does this).

## Manual partition assignment

... TODO ...

## Random ideas

A partition can be paused and un-paused every 2 cycles of the poll loop.
When this observed twice, we should automatically delay the pause by 1 poll loop.

Update: This idea has been implemented in https://github.com/zio/zio-kafka/pull/809.

**Comment by erikvanoosten** [comment]
PR #744 simplifies the runloop so that we can start working on problem no. 2.

**Comment by erikvanoosten** [comment]
This PR has been superseded by many other PRs. Most of the ideas have landed in zio-kafka 2.7.1. Closing for now.
